cEP-0018.md

```diff
@@ -0,0 +1,199 @@
+        C : [coantparsers.clexer, coantparsers.cparser],
+        Python : [coantparsers.pylexer, coantparsers.pyparser],
+        JavaScript : [coantparsers.jslexer, coantparsers.jsparser],
+        ...
+    }
+
+    @staticmethod
+    def loadFile(file,filename,lang = 'auto'):
+        """
+        Loads file's AST into memory
+        """
+        if(lang == 'auto'):
+          ext = get_extension(filename)
+        else:
+          ext = Languages(parse_lang_str(lang)[0])[0]
+        if ext is None:
+            raise FileExtensionNotSupported('Unknown File Type')
+        inputStream = file.readlines()
+        lexer = mapping[ext][0](inputStream)
+        parser = mapping[ext][1](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+
+    def get_imports():
+        """
+        Return list of strings containing import statements from the file
+        Requires the LANGUAGE parameter to contain exactly one language
+        """
+
+    def get_methods():
+        """
+        Return list of strings containing all methods from the file
+        Requires the LANGUAGE parameter to contain exactly one language
+        """
+    ...
+```
+
+### Prototype of  `ASTBear` class implementation
+
+```python
+from coantlib import ASTWalker
+from coalib.bears.LocalBear import Bear
+
+class ASTBear(LocalBear):
+    walker = None
+
+    def initialise(file, filename):
+        if LANGUAGES:
+            self.walker = ASTWalker(file, filename, self.LANGUAGES[0])
+        else:
+            self.walker = ASTWalker(file, filename, 'auto')
+
+    def run(self,
+            filename,
+            file,
+            tree,
+            lang = 'auto',
+            *args,
+            dependency_results = None,
+            **kwargs
+            ):
+        raise NotImplementedError('Needs to be done by bear')
+```
+
+### A test bear
+
+```python
+from coantlib.ASTBear import ASTBear
+from coalib.results.Result import Result
+
+class TestBear(ASTBear):
+    LANGUAGES = { 'Python 2' }
+    def run(self,
+            filename,
+            file,
+            tree,
+            lang = 'auto',
+            *args,
+            dependency_results=None,
+            **kwargs
+            ):
+        self.initialise(file, filename, lang)
+        num_nodes = 0
+        while(tree.get_next_node() != None):
```

Comment:
```
method_names = self.get_method_names()
for mn in method_names:
   # is_invalid_pep8_method should be implement coala developer not our side.
   # however get_method_names should be provided by our side(PythonASTBear)
    if not self.is_invalid_pep8_method(mn):
         yield Result('blah blah blah')
```

Reply:
@virresh
In my guess this comment will help you to understand what I want to say.